Multiple CH server setup

chumaumenze opened the pull request:
Closes #28

poundifdef commented on ch/server.go:
```diff
@@ -0,0 +1,90 @@
+		Config     ClickhouseConfig `mapstructure:"config"`
+
+		clickhouse.Conn
+		mu sync.Mutex
```
Curious, why do we need this?

poundifdef commented on ch/server.go:
```diff
@@ -0,0 +1,90 @@
+
+	ClickhouseServers struct {
+		Servers map[string]ClickhouseServer `mapstructure:"servers"`
+		mu      sync.Mutex
```
Same as above - does not look like this is used.

chumaumenze replied:
Exclusion lock. This is currently not used. I will remove it if necessary.

poundifdef commented on importer/importer.go:
```diff
@@ -369,61 +366,22 @@ func (im *Importer) insertData(conn driver.Conn, bucket, key, db, table string,
 
-	conn, err := im.connect()
-	if err != nil {
+	defaultServer, ok := im.Config.Clickhouse.Servers["default"]
```
Rather than doing this, I'd like to pass an instance of `ClickhouseProvider` into the `Importer`. Then have the default implementation fetch data from the config file.

If we do that, we could create a single `ClickhouseProvider` instance in `main.go` and pass it to any modules that need it. It would also let us avoid another refactor of the importer since we will only be accessing Clickhouse via the interface.

chumaumenze replied:
@poundifdef  this has similar effect as we are still reading from the config file. The major difference is that `*Importer.Start()` doesn't take the `ClickhouseProvider` parameter because `importer.NewImporter(&C)` already takes the config which includes the servers.

I can modify the `Start()` function to become `*Importer.Start(provider ClickhouseProvider)` and read in `Start(C.Clickhouse)`.

poundifdef replied:
Let's do that - I'd suggest adding it as a parameter to `NewImporter()`.

poundifdef reviewed (commented):
Awesome! A few comments but this is looking good.

chumaumenze commented:
@poundifdef  I refactored this. I also made a slight change to the cli flag. 

Execution would look like: `scratchdb -ingest -insert -config=config.toml`. Also, added context to properly handle signals and removed unnecessary wait groups.

poundifdef commented on main.go:
```diff
@@ -16,84 +19,64 @@ import (
 
 func main() {
-	log.SetFlags(log.LstdFlags | log.Lshortfile)
+	go func() {
```
Perhaps we can add a new flag, like `-pprof` for running this

chumaumenze replied:
Hehe... I forgot to clean up. Let me fix that.

chumaumenze replied:
We can have a different main target for it, separate files and all.

poundifdef commented:
Thank you, I'll pull this down and test locally.

poundifdef commented:
@chumaumenze Can you put the flag refactor as one PR, and the signal handling as another? That way we can merge smaller changes incrementally and I can review and test your original change (multiple clickhouse servers) in isolation.
